### DISCUSSION TITLE
Getting started with setup of a Rust3DS dev environment

### DISCUSSION THREAD

**Discussion by ian-h-chamberlain** [rust3ds/ctru-rs#80] [Q&A]
Spinning out into a discussion Q&A from some comments in #32 to avoid too much unrelated stuff happening on that issue. 

Context below:

> Hi, I just found this project, I dont have much experience in Rust, and also just started my into 3DS cfw development.
> Is there documentation on how to setup the rust 3ds dev environment? I have been trying, but there really no much documentation about it, and I have been unsuccessful.
----
> @PeCoBe I don’t know if I ever fully wrote it somewhere, but here it is:
> 
> - Make sure you’ve installed the devkitARM toolchain with all packages
> - Export these environment variables (put them in a .bashrc or similar):
> ```bash
> export DEVKITPRO=/opt/devkitpro
> export DEVKITARM=$DEVKITPRO/devkitARM
> 
> ```
> - Install the latest Rust nightly version (and use it for all further operations)
> - Install our `cargo-3ds` tool (using `cargo install`)
> - Now, you can either make your own project or run one of our examples in `ctru-rs`. Just use the command `cargo 3ds run`. For running an example, do `cargo 3ds run —example [example name]`
> 
> Without some knowledge of the Rust toolchain (or the devkitPRO one) it might be difficult to solve any issues, but this is more or less what you need to do.
> 
> Also, if you have little knowledge about writing custom homebrew, I suggest using the standard C toolchain by devkitPRO and play around with it. Our toolchain is far from finished or stable.
----

> Thanks. Yeah Im want to start with devkitpro first, but found this I tought hey lets give it a try.
> I think I was able to setup all, but having this error 
>  `   Compiling ctru-sys v0.4.1 (D:\Documents\Git\Rust3DS\ctru-rs\ctru-sys)
> error: could not find native static library `ctrud`, perhaps an -L flag is missing?`
> 
> anyway, thanks, I will keep an eye on this project.

**Comment by ian-h-chamberlain** [comment]
@PeCoBe that's interesting, perhaps your devkitPro installation is missing the debug version of the library? As another test, you could see if building with `--release` works, since that should link against `libctru` instead of `libctrud`.

I'm running macOS so the installation directory is probably different but on my system there are basically two copies of each library installed in `$DEVKITPRO/libctru/lib` (one debug and one release):
```
§ ls /opt/devkitpro/libctru/lib/
total 15096
drwxr-xr-x  8 root  wheel   256B Jun 30 21:13 ./
drwxr-xr-x  5 root  wheel   160B Jun 30 21:13 ../
-rw-r--r--  1 root  wheel   195K Jan 28  2022 libcitro2d.a
-rw-r--r--  1 root  wheel   185K Jan 28  2022 libcitro2dd.a
-rw-r--r--  1 root  wheel   663K Jan 28  2022 libcitro3d.a
-rw-r--r--  1 root  wheel   660K Jan 28  2022 libcitro3dd.a
-rw-r--r--  1 root  wheel   3.0M May 24 14:58 libctru.a
-rw-r--r--  1 root  wheel   2.7M May 24 14:58 libctrud.a
```

Does your installation look like that as well?

**Reply by PeCoBe** [reply]
Yeah looks the same.


```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a---         1/28/2022   5:23 PM         199708   libcitro2d.a
-a---         1/28/2022   5:23 PM         189468   libcitro2dd.a
-a---         1/28/2022   5:02 PM         678998   libcitro3d.a
-a---         1/28/2022   5:02 PM         675894   libcitro3dd.a
-a---         1/26/2022   5:17 PM        3139894   libctru.a
-a---         1/26/2022   5:17 PM        2859014   libctrud.a
```

I tried with --release got a similar error but now with ctru
```
   Compiling pthread-3ds v0.1.0 (https://github.com/rust3ds/pthread-3ds.git#c132988a)
error: could not find native static library `ctru`, perhaps an -L flag is missing?
```

**Reply by ian-h-chamberlain** [reply]
Hmm, in that case perhaps the right `RUSTFLAGS` are not being set for your environment. I wonder if this format doesn't work for directory separator reasons?
https://github.com/rust3ds/cargo-3ds/blob/master/src/lib.rs#L107-L111

You could try setting `RUSTFLAGS="-L \path\to\libctru\lib"` manually when running to see if that helps — if so, perhaps we need to account for directory separator properly in `cargo-3ds` to make it work...

**Reply by PeCoBe** [reply]
I was able to make it work, 
a brief of all that happened:
OS: Windows 10, PowerShell
Devkit was not configured correctly in my PATH, was using /opt directory but I was trying to run in from PowerShell, so had to update to the correct D:/ format. 
Probably because of this change, the linker `arm-none-eabi-gcc` was not found, tried installing it but then Techie told me the linker should already be under Devkitpro, was able to find it, added it to the PATH, and that fixed that error.
Cargo was able to compile, but then showed an error 'smdhtool' not found. Was able to locate it again under devkitpro folder, added the PATH on the same script used to add the linker path, and finally was able to run and create the .3ds file. 
Tested it on my 2DS, all working good.

**Reply by Meziu** [reply]
Yes, I told you it probably had to do with pathing. Glad to see it fixed, I hope you’ll find our work useful!
